### PR TITLE
(feat): support for non-default microphone input (by deviceId).

### PR DESCRIPTION
**Pull request by Domratchev**
I've given up on installing the package from the forked repo due to the issues with the prepack script.
See here for details: https://github.com/npm/npm/issues/19564

### CODE REVIEW

**Comment by wolfma61** [comment]
Thank you Alexei - we will take a look ... 
for any problems with the fork, please don't hesitate to file a separate issue

thx
Wolfgang

File: src/common.browser/MicAudioSource.ts [hunk 1]
```diff
@@ -45,16 +45,16 @@ export class MicAudioSource implements IAudioSource {
...
-    private privRecorder: IRecorder;
-
     private privMediaStream: MediaStream;
 
     private privContext: AudioContext;
 
-    public constructor(recorder: IRecorder, audioSourceId?: string) {
+    private readonly privConstraints: MediaStreamConstraints;
+
+    public constructor(private readonly privRecorder: IRecorder, constraints?: MediaStreamConstraints, audioSourceId?: string) {
```
[the remark this answers is not in the document]
Comment (Domratchev) [reply src/common.browser/MicAudioSource.ts]: It's a shorthand notation with exactly the same meaning as before. Private property is still there (also declared as readonly, which I would advise for a few other properties as well).

[the remark this answers is not in the document]
Comment (Domratchev) [reply src/common.browser/MicAudioSource.ts]: MediaStreamConstraints is just an interface with three properties. I only needed deviceId constraint on audio input, but then again others might want that to be exact or ideal, or specify groupId instead. So the idea was to give the user the complete choice.
I doubt there would be much value in MicAudioSource outside of browser anyway, as it solely relies on window.navigator implementation of MediaDevices Web API.

File: src/sdk/Audio/AudioConfig.ts [hunk 2]
```diff
@@ -24,6 +24,19 @@ export abstract class AudioConfig {
...
 
+    /**
+     * Creates an AudioConfig object representing a microphone on the system based on the specified constraints.
+     * @member AudioConfig.fromMicrophoneInput
+     * @function
+     * @public
+     * @param {MediaStreamConstraints} constraints A MediaStreamConstraints object specifying the requirements for microphone media device.
+     * @returns {AudioConfig} The audio input configuration being created.
+     */
+    public static fromMicrophoneInput(constraints?: MediaStreamConstraints): AudioConfig {
```
[the remark this answers is not in the document]
Comment (Domratchev) [reply src/sdk/Audio/AudioConfig.ts]: We could, of course, re-export MediaStreamConstraints, MediaTrackConstraints, MediaTrackConstraintSet, ConstrainBooleanParameters, ConstrainDoubleRange, ConstrainLongRange and ConstrainDOMStringParameters. But they are just POJO interfaces readily available in typescript/lib/lib.dom.d.ts, with the whole MicAudioSource class only working in browser (or with some very advanced polyfill for NodeJS).

File: .gitignore [hunk 3]
```diff
@@ -10,3 +10,4 @@ secrets/*
 test-javascript-junit.xml
 coverage/*
 *.tgz
+.vscode/settings.json
```
[the remark this answers is not in the document]
Comment (Domratchev) [reply .gitignore]: Done.

**Comment by wolfma61** [comment]
Thank you - this looks good. We are doing similar work in the SpeechSDK for the other languages. We want to make sure it is matching on the API level as closely as possible across all languages. 
Once we have the API solid, we should work together to adapt this PR here and merge, so making it a feature in the next release later this month.
It shouldn't delay it too much ...
Wolfgang

**Comment by Domratchev** [comment]
Any plans on merging it in any tine soon, please?

**Comment by rhurey** [comment]
Alexei,

First off, sorry about the delay in getting back to you.

We’ve been looking at what the right microphone search / enumeration / specification story is across all of the languages and platforms the speech SDK runs on, but have had a few competing priorities to juggle also.

While our long term aspirations are to have a complete story, we recognize the need for something now.

The proposal we’ve got now is to have a method fromMicrophoneDevice(…) that on JavaScript will take a MediaStream object and change MicAudioSource to take that as an additional constructor. 

Thoughts?

**Comment by Domratchev** [comment]
Ryan,

That is definitely possible, but somewhat complicated: MicAudioSource.turnOn() method would have to be significantly reworked then. Taking it even further, a new MediaStreamSource could be created along with fromMediaStream() helper method.
Yet all that sounds rather over-engineered, especially considering how small the original proposed change is. MicAudioSource class is already tightly coupled [internally] with WebRTC API. I don't think that a minor leak of this coupling to the public API could hurt. I would even be happy with just deviceId param for fromMicrophoneInput() method.

Cheers,
Alexei

**Comment by rhurey** [comment]
/**
     * Creates an audioConfig based from the microphone whose Id is a closes match
     * to the specified name.
     * @param deviceName The deviceId for the microphone to be returned.
     */
    public static fromMicrophoneInput(deviceName: string):AudioConfig{

**Comment by Domratchev** [comment]
Ryan,

It cannot be deviceName semantically, as the only [relevant] constraint supported by all browsers is deviceId: https://developer.mozilla.org/en-US/docs/Web/API/MediaTrackConstraints
Otherwise, passing just a deviceId in would be a satisfactory compromise.

Thanks,
Alexei

**Comment by Domratchev** [comment]
@rhurey Done.

**Pull request merged**